src/odemis/gui/comp/viewport.py

```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
+            self.bottom_legend.hi_ellipsis = self.canvas.display_xrange[1] < self.canvas.data_xrange[1]
+        self.bottom_legend.Refresh()
+
+        logging.debug("HRange: %s", hrange)
+        return hrange
+
+    def update_vrange(self, vrange):
+        """
+        Setter for VA vrange
+
+        Prevents VA change if the scale is locked.
+        Refreshes the plot and axis legend displays with the new scale
+        """
+        self.canvas.set_ranges(self.hrange.value, vrange)
+
+        self.left_legend.range = vrange
+        if self.canvas.has_data():
+            self.left_legend.lo_ellipsis = self.canvas.display_yrange[0] > self.canvas.data_yrange[0]
+            self.left_legend.hi_ellipsis = self.canvas.display_yrange[1] < self.canvas.data_yrange[1]
+        self.left_legend.Refresh()
+
+        logging.debug("VRange: %s", vrange)
+        return vrange
+
+    def on_hlegend_scroll(self, evt=None):
+        """ Scroll event for the bottom legend.
+
+        Zooms the x-scale of the data and refreshes display
+        """
+        if not self.canvas.has_data():
+            return
+
+        rot = evt.GetWheelRotation() / evt.GetWheelDelta()
+        span = self.hrange.value[1] - self.hrange.value[0]
+        zoom_centre = self.canvas.pos_x_to_val_x(evt.Position[0])
+        scale = 0.1
+        rng = self.hrange.value
+
+        # proportion of zoom_centre
+        prop = (zoom_centre - rng[0]) / span
+        
+        # zoom around the centre point
+        if rot < 0:
+            new_span = span * (1 + scale)
+        else:
+            new_span = span * (1 - scale)
+
+        lo = zoom_centre - prop * new_span
+        hi = lo + new_span
+
+        # Clamp the ranges - this functions differently for scrolls than pans
+        if lo < self.canvas.data_xrange[0]:
+            lo = self.canvas.data_xrange[0]
+        if hi > self.canvas.data_xrange[1]:
+            hi = self.canvas.data_xrange[1]
+
+        self.hrange.value = (lo, hi)
+
+    def on_vlegend_scroll(self, evt=None):
+        """ Scroll event for the left legend.
+
+        Zooms the y-scale of the data and refreshes display
+        """
+        if not self.canvas.has_data():
+            return
+
+        rot = evt.GetWheelRotation() / evt.GetWheelDelta()
+        span = self.vrange.value[1] - self.vrange.value[0]
+        # zoom_centre = self.canvas.pos_x_to_val_x(evt.Position[0])
+        zoom_centre = self.canvas.pos_y_to_val_y(evt.Position[1])
+        scale = 0.1
+        rng = self.vrange.value
+
+        # proportion of zoom_centre
+        prop = (zoom_centre - rng[0]) / span
+
+        # zoom around the centre point
+        if rot < 0:
+            new_span = span * (1 + scale)
+        else:
+            new_span = span * (1 - scale)
+
+        lo = zoom_centre - prop * new_span
+        hi = lo + new_span
+
+        # Clamp the ranges - this functions differently for scrolls than pans
+        if lo < self.canvas.data_yrange[0]:
```

Comment:
For such simple clip/clamp, you can just write like:
```
lo = max(self.canvas.data_yrange[0], lo)
hi = min(hi, self.canvas.data_yrange[1] + self.canvas.data_yrange[1] * self._vmargin)
```

BTW, be aware that this margin computation works only because yrange[0] = 0 (and yrange[1] is always > 0). I think the complete way to write it is `self.canvas.data_yrange[1] + (self.canvas.data_yrange[1] - self.canvas.data_yrange[0]) * self._vmargin`.